### PR TITLE
Array property type creation

### PR DESCRIPTION
**Pull request by luisbettencourt**
## 🌟 What is the purpose of this PR?

This PR focuses on implementing the logic for creating array property types.

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1203301853897838/1203366629132085/f) _(internal)_

## 🔍 What does this change?

- Added menus for creating a custom property type (only arrays in this PR);
- Moved some of the form logic on `property-type-form` to `property-utils`;
- Moved the dropdown on `property-type-selector-dropdown` to a generic component so it can be reused.

## 🚫 Blocked by

- [x]  #1527 
- [x]  #1521

## 🐾 Next steps

- Fix Modal positioning;
- Finish styling;
- Other subtasks [here](https://app.asana.com/0/1203301853897838/1203343922163681/f).

## ❓ How to test this?

1.  Start creating an entity type;
2. Add a custom property type;
3. Add an array (and nested properties) to the custom property type;
4. Save the property type;
5. Check if it was correctly saved in the database.

### CODE REVIEW

**Comment by nathggns** [comment]
Reviewing this now

File: packages/hash/frontend/src/pages/[account-slug]/types/entity-type/property-type-utils.ts [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+        },
+        minItems: data?.minItems,
+        maxItems: data?.maxItems,
+      };
+    }
+
+    return {
+      $ref: property!.data!.typeId,
+    };
+  }) as [PropertyValues, ...PropertyValues[]];
```
Comment (luisbettencourt) [inline packages/hash/frontend/src/pages/[account-slug]/types/entity-type/property-type-utils.ts:107]: We'll prevent expectedValues from being empty with validation in the future, but even returning early or throwing an error if the array is empty didn't satisfy typescript, so I had to cast this.

File: packages/hash/frontend/src/pages/[account-slug]/types/entity-type/property-type-utils.ts [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    icon: faList.icon,
+  },
+  [types.dataType.object.dataTypeId]: {
+    title: "Property Object",
+    icon: faCube,
+  },
+};
+
+export const getDefaultData = (
+  typeId: string,
```
Comment (nathggns) [inline packages/hash/frontend/src/pages/[account-slug]/types/entity-type/property-type-utils.ts:68]: Can this be `VersionedUri | "array"` to remove the need to cast?

**Review by nathggns — COMMENTED** [review]
I haven't had time to functionality review – but just some quick code comments. I think you should get a more thorough review before merging. 

Well done Luís!

File: packages/hash/frontend/src/pages/[account-slug]/types/entity-type/array-property-type-menu.tsx [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+    name: `flattenedPropertyList`,
+  });
+
+  const expectedValues = useWatch({
+    control,
+    name: `flattenedPropertyList.${id}.data.expectedValues`,
+  });
+
+  const dataTypeOptions = useMemo(
+    () => [...primitiveDataTypeOptions, ...customDataTypeOptions],
```
[the remark this answers is not in the document]
Comment (luisbettencourt) [reply packages/hash/frontend/src/pages/[account-slug]/types/entity-type/array-property-type-menu.tsx]: Done in d3304839812081067a92e356e5d75e35c1449dbb.

File: packages/hash/frontend/src/pages/[account-slug]/types/entity-type/array-property-type-menu.tsx [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+      >
+        {expectedValues?.map((childId, pos) => {
+          const property = flattenedProperties[childId];
+
+          if (!property?.data) {
+            return null;
+          }
+
+          return property.data.typeId === "array" ? (
+            <ArrayPropertyTypeMenu id={childId} index={[...index, pos]} />
```
[the remark this answers is not in the document]
Comment (luisbettencourt) [reply packages/hash/frontend/src/pages/[account-slug]/types/entity-type/array-property-type-menu.tsx]: Addressed in e57543c71edc4ff1701e8383ea4178fb9eb95373.

File: packages/hash/frontend/src/pages/[account-slug]/types/entity-type/custom-property-type-menu.tsx [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+            border: `1px solid ${palette.gray[30]}`,
+            borderBottomRightRadius: 4,
+            borderBottomLeftRadius: 4,
+          })}
+        >
+          <Button
+            size="small"
+            onClick={() => {
+              setValue(`expectedValues`, [
+                ...expectedValues,
```
[the remark this answers is not in the document]
Comment (luisbettencourt) [reply packages/hash/frontend/src/pages/[account-slug]/types/entity-type/custom-property-type-menu.tsx]: Changed to `getValues` in b782eb41f090758ccf81ce504ca7b9b84b6d74bd

File: packages/hash/frontend/src/pages/[account-slug]/types/entity-type/property-menu.tsx [hunk 6]
```diff
@@ -27,6 +27,7 @@ import {
   OntologyChip,
   parseUriForOntologyChip,
 } from "../../shared/ontology-chip";
+import { withHandler } from "./util";
```
[the remark this answers is not in the document]
Comment (luisbettencourt) [reply packages/hash/frontend/src/pages/[account-slug]/types/entity-type/property-menu.tsx]: It was a residual from a previous merge. Removed in dca170b8c9cf1bedcd8b70f30bbe17b37aef1f24.

File: packages/hash/frontend/src/pages/[account-slug]/types/entity-type/property-type-utils.ts (hunk 2 above)
Comment (luisbettencourt) [reply packages/hash/frontend/src/pages/[account-slug]/types/entity-type/property-type-utils.ts]: Done in 53b318c6cdd9de7817f0149e0ff3f0e439d2f91b. Thank you

File: packages/hash/frontend/src/pages/[account-slug]/types/entity-type/array-property-type-menu.tsx [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+          flex: 1,
+          background: ({ palette }) =>
+            palette.gray[index.length % 2 !== 0 ? 20 : 10],
+          borderBottomRightRadius: 4,
+          borderBottomLeftRadius: 4,
+          position: "relative",
+        }}
+      >
+        {expectedValues?.map((childId, pos) => {
+          const property = flattenedProperties[childId];
```
[the remark this answers is not in the document]
Comment (luisbettencourt) [reply packages/hash/frontend/src/pages/[account-slug]/types/entity-type/array-property-type-menu.tsx:97]: I'm converting this to a component in #1513, so I would rather address this in that PR. Thank you for the suggestion.

File: packages/hash/frontend/src/pages/[account-slug]/types/entity-type/custom-property-type-menu.tsx [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+                    fontWeight: 500,
+                    color: ({ palette }) => palette.gray[60],
+                  }}
+                >
+                  ALLOWS COMBINING
+                </Typography>
+                <Stack direction="row" gap={1}>
+                  <CustomChip color="purple" label="PROPERTY OBJECTS" />
+                  <CustomChip color="blue" label="DATA TYPES" />
+                  <CustomChip color="navy" label="ARRAY" />
```
Comment (CiaranMn) [inline packages/hash/frontend/src/pages/[account-slug]/types/entity-type/custom-property-type-menu.tsx:179]:
```suggestion
                  <CustomChip color="navy" label="ARRAYS" />
```

File: packages/hash/frontend/src/pages/[account-slug]/types/entity-type/custom-property-type-menu.tsx [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+              fontSize: 12,
+              color: ({ palette }) => palette.gray[50],
+              [`.${buttonClasses.endIcon}`]: {
+                ml: 0,
+              },
+            }}
+            endIcon={<FontAwesomeIcon icon={faClose} />}
+            variant="tertiary_quiet"
+          >
+            CANCEL
```
Comment (CiaranMn) [inline packages/hash/frontend/src/pages/[account-slug]/types/entity-type/custom-property-type-menu.tsx:88]: Can we put a bit of margin to the left of the CANCEL text? It looks a bit crowded at the moment

**Current**
Screenshot 2022-12-02 at 08 44 22

**Design**
Screenshot 2022-12-02 at 08 44 15

**Review by CiaranMn — COMMENTED** [review]
Had a test and noticed a couple of small things – looks good to me otherwise. 

Let me know when you're finished with Nate's feedback and I'll approve

File: packages/hash/frontend/src/pages/[account-slug]/types/entity-type/custom-property-type-menu.tsx (hunk 8 above)
Comment (luisbettencourt) [reply packages/hash/frontend/src/pages/[account-slug]/types/entity-type/custom-property-type-menu.tsx]: Fixed in 143ecff2f511740c90060e7e4a5c7d6c3e24a0a0. Well spotted

File: packages/hash/frontend/src/pages/[account-slug]/types/entity-type/custom-property-type-menu.tsx (hunk 9 above)
Comment (luisbettencourt) [reply packages/hash/frontend/src/pages/[account-slug]/types/entity-type/custom-property-type-menu.tsx:92]: Made several changes to this button in dd9f15bc762278ef8bad137dd33f88cfc2997dc6. Thank you.

**Review by CiaranMn — APPROVED** [review]

**Pull request merged**